v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/transformer/SourceRowToMutationDoFn.java

```diff
@@ -0,0 +1,87 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.transformer;
+
+import com.google.auto.value.AutoValue;
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.Value;
+import com.google.cloud.teleport.v2.source.reader.io.row.SourceRow;
+import com.google.cloud.teleport.v2.source.reader.io.schema.SourceTableReference;
+import com.google.cloud.teleport.v2.spanner.migrations.avro.GenericRecordTypeConvertor;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.ISchemaMapper;
+import java.io.Serializable;
+import java.util.Map;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is a DoFn class that takes a {@link SourceRow} and converts it into a {@link
+ * com.google.cloud.spanner.Mutation}.
+ */
+@AutoValue
+public abstract class SourceRowToMutationDoFn extends DoFn<SourceRow, Mutation>
+    implements Serializable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SourceRowToMutationDoFn.class);
+
+  public abstract ISchemaMapper iSchemaMapper();
+
+  public abstract Map<String, SourceTableReference> tableIdMapper();
+
+  public static SourceRowToMutationDoFn create(
+      ISchemaMapper iSchemaMapper, Map<String, SourceTableReference> tableIdMapper) {
+    return new AutoValue_SourceRowToMutationDoFn(iSchemaMapper, tableIdMapper);
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c) {
+    SourceRow sourceRow = c.element();
+    GenericRecord record = sourceRow.getPayload();
+    if (!tableIdMapper().containsKey(sourceRow.tableSchemaUUID())) {
+      LOG.error(
+          "cannot find valid sourceTable for tableId: {} in tableIdMapper",
+          sourceRow.tableSchemaUUID());
+      return;
+    }
+    String srcTableName = tableIdMapper().get(sourceRow.tableSchemaUUID()).sourceTableName();
+    try {
+      // TODO: update namespace in constructor when Spanner namespace support is added.
+      GenericRecordTypeConvertor genericRecordTypeConvertor =
+          new GenericRecordTypeConvertor(iSchemaMapper(), "");
+      Map<String, Value> values =
+          genericRecordTypeConvertor.transformChangeEvent(record, srcTableName);
+      String spannerTableName = iSchemaMapper().getSpannerTableName("", srcTableName);
+      Mutation mutation = mutationFromMap(spannerTableName, values);
+      c.output(mutation);
+    } catch (Exception e) {
+      // TODO: Add DLQ integration once supported.
+      LOG.error(
+          "Unable to transform source row to spanner mutation: {} {}",
+          e.getMessage(),
+          e.fillInStackTrace());
+    }
+  }
+
+  static Mutation mutationFromMap(String spannerTableName, Map<String, Value> values) {
```

Comment:
I've seen this in other PRs as well - Is there a reason you do not add `private/public` access modifiers for methods? I think it's almost a convention to do that, right? 
For e.g `mutationFromMap` should be private IMO?

Reply:
Yes I always thought not specifying defaults to private, just realised it is a separate access specifier. Will add :P

Reply:
Hehe, thanks. Please check other places as well once!